File: src/engine/executor.ts
import { EventEmitter } from 'events';
import { spawn, type ChildProcess } from 'child_process';
import treeKill from 'tree-kill';
import type { ApprovalMode, Backend } from '../shared/protocol.js';
import { createStrategy } from './strategy-factory.js';
import type { CliStrategy, OutputHandler } from './strategy.js';

export interface ExecutorOpts {
  prompt: string;
  workingDir: string;
  model?: string;
  approvalMode: ApprovalMode;
  timeout: number;
  backend: Backend;
}

/**
 * Wraps a single CLI process.
 * Uses strategy pattern to support different backends (Gemini, Kimi, etc.)
 */
export class CliExecutor extends EventEmitter {
  private opts: ExecutorOpts;
  private strategy: CliStrategy;
  private handler: OutputHandler;
  private proc: ChildProcess | null = null;
  private timer: NodeJS.Timeout | null = null;
  private alive = false;

  constructor(opts: ExecutorOpts) {
    super();
    this.opts = opts;
    this.strategy = createStrategy(opts.backend);
    this.handler = this.strategy.createOutputHandler(this);
  }

  get sessionId(): string | null {
    // Get from handler if available
    return null;
  }

  get pid(): number | undefined {
    return this.proc?.pid;
  }

  get isAlive(): boolean {
    return this.alive;
  }

  launch(): void {
    if (this.alive) throw new Error('Executor already running');
    this.alive = true;

    const { cmd, args, useStdin } = this.strategy.buildCommand(this.opts);

    this.proc = spawn(cmd, args, {
      cwd: this.opts.workingDir,
      stdio: ['pipe', 'pipe', 'pipe'],
    });

    // Feed prompt via stdin if needed
    if (useStdin && this.proc.stdin) {
      this.proc.stdin.write(this.opts.prompt);
      this.proc.stdin.end();
    }

    // Handle output using strategy
    if (this.proc.stdout) {
      this.handler.handleOutput(this.proc.stdout);
    }

    this.proc.on('exit', (code) => {
      this.teardown();
      this.emit('exit', code);
    });

    this.proc.on('error', (err) => {
      this.teardown();
      this.emit('error', err);
    });

    // Arm timeout
    if (this.opts.timeout > 0) {
      this.timer = setTimeout(() => {
        this.emit('timeout');
        this.kill();
      }, this.opts.timeout * 1000);
    }
  }

  kill(force = false): void {
    if (!this.proc?.pid || !this.alive) return;

    const sig = force ? 'SIGKILL' : 'SIGTERM';
    treeKill(this.proc.pid, sig, (err) => {
      if (err && this.alive) {
        try { this.proc?.kill(sig); } catch { /* already dead */ }
      }
    });

    // Escalate after 5s
    if (!force) {
      setTimeout(() => {
        if (this.proc?.pid && this.alive) {
          treeKill(this.proc.pid, 'SIGKILL');
        }
      }, 5000);
    }
  }

  private teardown(): void {
    this.alive = false;
    this.handler.destroy();
    if (this.timer) {
      clearTimeout(this.timer);
      this.timer = null;
    }
    this.proc = null;
  }
}

// Legacy alias for backward compatibility
export const GeminiExecutor = CliExecutor;

File: engine/executor.go
package engine

import (
	"cli-agent-go/engine/strategy"
	"context"
	"fmt"
	"io"
	"os"
	"os/exec"
	"sync"
	"syscall"
	"time"
)

// ExecutorEvent 执行器事件
type ExecutorEvent struct {
	Type string
	Data interface{}
}

// SimpleEventEmitter 简单的事件发射器实现
type SimpleEventEmitter struct {
	handlers map[string][]func(interface{})
	mu       sync.RWMutex
}

// NewSimpleEventEmitter 创建事件发射器
func NewSimpleEventEmitter() *SimpleEventEmitter {
	return &SimpleEventEmitter{
		handlers: make(map[string][]func(interface{})),
	}
}

// Emit 发射事件
func (e *SimpleEventEmitter) Emit(event string, data interface{}) {
	e.mu.RLock()
	handlers := e.handlers[event]
	e.mu.RUnlock()
	for _, h := range handlers {
		go h(data)
	}
}

// On 注册事件处理器
func (e *SimpleEventEmitter) On(event string, handler func(interface{})) {
	e.mu.Lock()
	defer e.mu.Unlock()
	e.handlers[event] = append(e.handlers[event], handler)
}

// CliExecutor 包装单个 CLI 进程
type CliExecutor struct {
	opts       strategy.ExecutorOpts
	strategy   strategy.CliStrategy
	handler    strategy.OutputHandler
	emitter    *SimpleEventEmitter
	proc       *exec.Cmd
	cancelFunc context.CancelFunc
	alive      bool
	mu         sync.Mutex
	timer      *time.Timer
}

// NewCliExecutor 创建 CLI 执行器
func NewCliExecutor(opts strategy.ExecutorOpts) *CliExecutor {
	s := strategy.CreateStrategy(opts.Backend)
	emitter := NewSimpleEventEmitter()
	handler := s.CreateOutputHandler(emitter)

	return &CliExecutor{
		opts:     opts,
		strategy: s,
		handler:  handler,
		emitter:  emitter,
		alive:    false,
	}
}

// SessionID 获取会话 ID
func (e *CliExecutor) SessionID() string {
	return ""
}

// PID 获取进程 ID
func (e *CliExecutor) PID() int {
	if e.proc != nil && e.proc.Process != nil {
		return e.proc.Process.Pid
	}
	return 0
}

// IsAlive 检查是否存活
func (e *CliExecutor) IsAlive() bool {
	e.mu.Lock()
	defer e.mu.Unlock()
	return e.alive
}

// On 注册事件处理器
func (e *CliExecutor) On(event string, handler func(interface{})) {
	e.emitter.On(event, handler)
}

// Launch 启动执行器
func (e *CliExecutor) Launch() error {
	e.mu.Lock()
	defer e.mu.Unlock()

	if e.alive {
		return fmt.Errorf("executor already running")
	}
	e.alive = true

	cmdStr, args, useStdin := e.strategy.BuildCommand(e.opts)

	ctx, cancel := context.WithCancel(context.Background())
	e.cancelFunc = cancel

	e.proc = exec.CommandContext(ctx, cmdStr, args...)
	e.proc.Dir = e.opts.WorkingDir
	e.proc.SysProcAttr = &syscall.SysProcAttr{
		Setpgid: true,
	}

	// Feed prompt via stdin if needed
	if useStdin {
		e.proc.Stdin = stringsToReader(e.opts.Prompt)
	}

	// Get stdout pipe
	stdout, err := e.proc.StdoutPipe()
	if err != nil {
		e.alive = false
		return fmt.Errorf("failed to create stdout pipe: %w", err)
	}

	// Get stderr pipe
	stderr, err := e.proc.StderrPipe()
	if err != nil {
		e.alive = false
		return fmt.Errorf("failed to create stderr pipe: %w", err)
	}

	// Start the process
	if err := e.proc.Start(); err != nil {
		e.alive = false
		return fmt.Errorf("failed to start process: %w", err)
	}

	// Handle output using strategy
	go e.handler.HandleOutput(stdout)
	
	// Handle stderr
	go io.Copy(io.Discard, stderr)

	// Wait for process to exit
	go func() {
		err := e.proc.Wait()
		e.teardown()
		if err != nil {
			e.emitter.Emit("exit", 1)
		} else {
			e.emitter.Emit("exit", 0)
		}
	}()

	// Arm timeout
	if e.opts.Timeout > 0 {
		e.timer = time.AfterFunc(time.Duration(e.opts.Timeout)*time.Second, func() {
			e.emitter.Emit("timeout", nil)
			e.Kill(false)
		})
	}

	return nil
}

// Kill 杀死进程
func (e *CliExecutor) Kill(force bool) {
	e.mu.Lock()
	defer e.mu.Unlock()

	if e.proc == nil || e.proc.Process == nil || !e.alive {
		return
	}

	sig := syscall.SIGTERM
	if force {
		sig = syscall.SIGKILL
	}

	// Try to kill the process group
	syscall.Kill(-e.proc.Process.Pid, sig)

	// If not force, escalate after 5s
	if !force {
		time.AfterFunc(5*time.Second, func() {
			e.mu.Lock()
			defer e.mu.Unlock()
			if e.proc != nil && e.proc.Process != nil && e.alive {
				syscall.Kill(-e.proc.Process.Pid, syscall.SIGKILL)
			}
		})
	}
}

func (e *CliExecutor) teardown() {
	e.mu.Lock()
	defer e.mu.Unlock()

	e.alive = false
	e.handler.Destroy()
	if e.timer != nil {
		e.timer.Stop()
		e.timer = nil
	}
	if e.cancelFunc != nil {
		e.cancelFunc()
	}
	e.proc = nil
}

func stringsToReader(s string) io.Reader {
	return &stringReader{s: s, i: 0}
}

type stringReader struct {
	s string
	i int
}

func (r *stringReader) Read(p []byte) (n int, err error) {
	if r.i >= len(r.s) {
		return 0, io.EOF
	}
	n = copy(p, r.s[r.i:])
	r.i += n
	return n, nil
}

// Legacy alias
var GeminiExecutor = NewCliExecutor

// MockOsStdin 用于测试
var MockOsStdin = os.Stdin
